src/common/async/redirect_error.h

```diff
@@ -0,0 +1,342 @@
+/**
+ * The redirect_error_t class is used to indicate that any disposition produced
+ * by an asynchronous operation is captured to a specified variable.
+ */
+template<typename CompletionToken,
+         boost::asio::disposition Disposition>
+class redirect_error_t {
+public: // Sigh
+  CompletionToken token;
+  Disposition& disposition;
+
+  template<typename CT>
+  redirect_error_t(CT&& token, Disposition& disposition)
+    : token(std::forward<CT>(token)), disposition(disposition) {}
+};
+
+/// A function object type that adapts a @ref completion_token to capture
+/// disposition values to a variable.
+/**
+ * May also be used directly as a completion token, in which case it adapts the
+ * asynchronous operation's default completion token (or boost::asio::deferred
+ * if no default is available).
+ */
+template<boost::asio::disposition Disposition>
+class partial_redirect_error {
+public:
+  Disposition& disposition;
+
+  /// Constructor that specifies the variable used to capture disposition values.
+  explicit partial_redirect_error(Disposition& disposition)
+    : disposition(disposition) {}
+
+  /// Adapt a @ref completion_token to specify that the completion handler
+  /// should capture disposition values to a variable.
+  template<typename CompletionToken>
+  [[nodiscard]] constexpr inline auto
+  operator ()(CompletionToken&& token) const {
+    return redirect_error_t<std::decay_t<CompletionToken>, Disposition>{
+      std::forward<CompletionToken>(token), disposition};
+  }
+};
+
+/// Create a partial completion token adapter that captures disposition values
+/// to a variable.
+template<boost::asio::disposition Disposition>
+[[nodiscard]] inline auto redirect_error(Disposition& d)
+{
+  return partial_redirect_error<std::decay_t<Disposition>>{d};
+}
+
+/// Adapt a @ref completion_token to capture disposition values to a variable.
+template<typename CompletionToken, boost::asio::disposition Disposition>
+[[nodiscard]] inline auto redirect_error(CompletionToken&& token,
+                                         Disposition& d)
+{
+  return redirect_error_t<std::decay_t<CompletionToken>,
+                          std::decay_t<Disposition>>{
+    std::forward<CompletionToken>(token), d};
+}
+
+namespace detail {
+template<typename Handler, boost::asio::disposition Disposition>
+class redirect_error_handler {
+public:
+  Disposition& disposition;
+  Handler handler;
+
+  using result_type = void;
+  template<typename CompletionToken>
+  redirect_error_handler(
+    redirect_error_t<std::decay_t<CompletionToken>,
+                     std::decay_t<Disposition>> re)
+    : disposition(re.disposition), handler(std::move(re.token)) {}
+
+  template<typename RedirectedHandler>
+  redirect_error_handler(Disposition &disposition, RedirectedHandler&& handler)
+    : disposition(disposition),
+      handler(std::forward<RedirectedHandler>(handler)) {}
+
+
+  void operator ()() {
+    std::move(handler)();
```

Comment:
*scratching head a bit*

Reply:
There are some ASIO-isms here I'm surely not familiar with-- it seems odd to move to something (in the ctor) that looks concrete and then before invoking move() the same concrete-looking thing to call operator(). Magic must be happening. It might be worth a comment, unless folks more familiar with ASIO would be reasonably expected to know what's up here.

Reply:
Should you find yourself in there, a comment to that effect might be helpful, but if it's a common/known asio idiom, disregard.